node/actors/bft/src/testonly/run.rs

```diff
@@ -71,77 +110,407 @@ impl Test {
+                            for target_port in target_ports {
+                                if p.contains(target_port) {
+                                    eprintln!("unicasting view={view_number} from={port} target={target_port } can_send={can_send} t={}", start.elapsed().as_secs());
+                                    send_or_stash(can_send, *target_port, msg());
                                 }
                             }
-                            Ok(())
-                        });
+                        }
                     }
-                    anyhow::Ok(())
-                })
-                .await?;
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Simulate the effects of gossip: if the source node can gossip to the target,
+/// and the message being sent contains a CommitQC, and the sender has the
+/// referenced finalized block in its store, then assume that the target
+/// could fetch this block if they wanted, and insert it directly into their store.
+///
+/// This happens concurrently with the actual message passing, so it's just an
+/// approximation. It also happens out of order. This method only contains the
+/// send loop, to deal with the spawning of store operations.
+async fn twins_gossip_loop(
+    ctx: &ctx::Ctx,
+    stores: &HashMap<Port, Arc<BlockStore>>,
+    mut recv: UnboundedReceiver<(Port, Port, BlockNumber)>,
+) -> anyhow::Result<()> {
+    scope::run!(ctx, |ctx, s| async move {
+        while let Ok((from, to, mut number)) = recv.recv(ctx).await {
+            let local_store = &stores[&from];
+            let remote_store = &stores[&to];
+            let first_needed = remote_store.queued().next();
+
+            loop {
+                // Stop going back if the target already has the block.
+                if number < first_needed {
+                    break;
+                }
+                // Stop if the source doesn't actually have this block to give.
+                let Ok(Some(block)) = local_store.block(ctx, number).await else {
+                    break;
+                };
+                // Perform the storing operation asynchronously because `queue_block` will
+                // wait for all dependencies to be inserted first.
+                s.spawn_bg(async move {
+                    let _ = remote_store.queue_block(ctx, block).await;
+                    Ok(())
+                });
+                // Be pessimistic and try to insert all ancestors, to minimise the chance that
+                // for some reason a node doesn't get a finalized block from anyone.
+                // Without this some scenarios with twins actually fail.
+                if let Some(prev) = number.prev() {
+                    number = prev;
+                } else {
+                    break;
+                };
             }
         }
         Ok(())
     })
     .await
 }
+
+fn output_msg_view_number(msg: &io::OutputMessage) -> u64 {
```

Comment:
afaict there is no need to strip the type safety here. Returning ViewNumber would be more readable, no?